src/registry.ts

```diff
@@ -0,0 +1,124 @@
+import { each, has, reduce, valuesIn } from 'lodash';
+import { Collector, CounterValue, HistogramValue, Metric } from './collector';
+import { Counter } from './counter';
+import { Gauge } from './gauge';
+import { Histogram } from './histogram';
+
+import { formatCounterOrGauge, formatHistogramOrSummary } from './utils';
+
+export type CollectorType = 'counter' | 'gauge' | 'histogram';
+
+type CollectorForType<T extends CollectorType> =
+  T extends 'histogram' ? Histogram :
+    T extends 'gauge' ? Gauge :
+      T extends 'counter' ? Counter :
+        never;
+
+type RegistryItem<T extends CollectorType> = {
+  [key: string]: {
+    type: T,
+    help: string,
+    instance: CollectorForType<T>,
+  },
+};
+
+export class Registry {
+  private data: {
+    [K in CollectorType]: RegistryItem<K>
+  };
+
+  constructor() {
+    this.data = {
+      counter: {},
+      gauge: {},
+      histogram: {},
+    };
+  }
+
+  private checkExists(type: CollectorType, name: string): void {
+    // check for name collisions
```

Comment:
might be best to just change the name to one that is obvious to not need a comment or use JSDoc https://github.com/Microsoft/TypeScript/wiki/JSDoc-support-in-JavaScript